### PR TITLE
drivers: rework usb_dc_mcux_ehci

### LINKED ISSUE #19763: tests/subsys/usb/device/ failed on mimxrt1050_evk board.
**To Reproduce**
1. west build -b mimxrt1050_evk tests/subsys/usb/device/
2. west flash
3. See error

**Screenshots or console output**
***** Booting Zephyr OS build zephyr-v2.0.0-1126-g91384ef01eb6 *****
D: __usb_descriptor_start 0x80002b68
D: __usb_descriptor_end 0x80002bc1
D: Configuration descriptor 0x80002b7a
D: Interface descriptor 0x80002b83
D: Endpoint descriptor 0x80002b8c
D: Fixing EP address 81 -> 81
D: endpoint 0x81
D: Now the wTotalLength is 25
D: idx_max 11, ascii_idx_max 5, buf 0x80002b99
D: char R : 52, idx 5 -> 11
D: char Y : 59, idx 4 -> 9
D: char H : 48, idx 3 -> 7
D: char P : 50, idx 2 -> 5
D: char E : 45, idx 1 -> 3
D: char Z : 5a, idx 0 -> 1
D: idx_max 13, ascii_idx_max 6, buf 0x80002ba7
D: char V : 56, idx 6 -> 13
D: char E : 45, idx 5 -> 11
D: char D : 44, idx 4 -> 9
D: char - : 2d, idx 3 -> 7
D: char B : 42, idx 2 -> 5
D: char S : 53, idx 1 -> 3
D: char U : 55, idx 0 -> 1
D: idx_max 7, ascii_idx_max 3, buf 0x80002bb7
D: char 1 : 31, idx 3 -> 7
D: char 0 : 30, idx 2 -> 5
D: char . : 2e, idx 1 -> 3
D: char 0 : 30, idx 0 -> 1
D: attached
D: set cb, ep: 0x81
Running test suite test_device

starting test - test_usb_dc_api_invalid

    Assertion failed at ZEPHYR_BASE/tests/subsys/usb/device/src/main.c:124: test_usb_dc_api_invalid: (usb_dc_ep_set_stall(INVALID_EP) equal to TC_PASS)
usb_dc_ep_set_stall(INVALID_EP)
FAIL - test_usb_dc_api_invalid

starting test - test_usb_dc_api
PASS - test_usb_dc_api

starting test - test_usb_dc_api_read_write
E: Wrong endpoint index/address/direction
E: ***** BUS FAULT *****
E:   Imprecise data bus error
E: r0/a1:  0x800012f8  r1/a2:  0x00000080  r2/a3:  0x40184000
E: r3/a4:  0x00000001 r12/ip:  0x00000000 r14/lr:  0x60004607
E:  xpsr:  0x61000000
E: Faulting instruction address (r15/pc): 0x600055cc
E: >>> ZEPHYR FATAL ERROR 0: CPU exception
E: Current thread: 0x8000136c (unknown)
E: Halting system

**Environment (please complete the following information):**
 - OS: Fedora29
 - Toolchain: zephyr-sdk-0.10.3
 - Commit ID: 91384ef01

### PR DESCRIPTION
**Pull request by jfischer-no**
This is bug fix, rework, and style fix of usb_dc_mcux_ehci

Fixes #19763

### CODE REVIEW

**Comment by jfischer-phytec-iot** [comment]
Tested on IMXRT1050-EVKB, passes USB 2 CV Chapter 9 and all testusb tests now.

**Comment by ioannisg** [comment]
We should perhaps hold this back, merge it after 2.1 release, we are in -rc3 already.

**Review by eobalski — CHANGES_REQUESTED** [review]

**Comment by MaureenHelm** [comment]
@MarkWangChinese please review

File: drivers/usb/device/usb_dc_mcux_ehci.c [hunk 1]
```diff
@@ -45,65 +51,93 @@ K_MEM_POOL_DEFINE(ep_buf_pool, 16, 1024, EP_BUF_NUMOF_BLOCKS, 4);
...
+		return -EINVAL;
+	}
+
+	status = dev_data.interface->deviceDeinit(dev_data.controllerHandle);
+	if (kStatus_USB_Success != status) {
+		return -EINVAL;
+	}
+
+	dev_data.controllerHandle = NULL;
+	dev_data.attached = 0;
```
[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply drivers/usb/device/usb_dc_mcux_ehci.c]: fixed

File: drivers/usb/device/usb_dc_mcux_ehci.c [hunk 2]
```diff
@@ -45,65 +51,93 @@ K_MEM_POOL_DEFINE(ep_buf_pool, 16, 1024, EP_BUF_NUMOF_BLOCKS, 4);
...
+		LOG_WRN("Device not attached");
+		return 0;
 	}
-	s_Device.attached = 0;
+
+	status = dev_data.interface->deviceControl(dev_data.controllerHandle,
+						   kUSB_DeviceControlStop,
+						   NULL);
+	if (kStatus_USB_Success != status) {
+		return -EINVAL;
```
[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply drivers/usb/device/usb_dc_mcux_ehci.c]: thanks for the point, forgot to fix it here.

File: drivers/usb/device/usb_dc_mcux_ehci.c [hunk 3]
```diff
@@ -139,79 +173,143 @@ int usb_dc_ep_check_cap(const struct usb_dc_ep_cfg_data *const cfg)
...
-		memset(block->data, 0, cfg->ep_mps);
-	} else {
+	if (k_mem_pool_alloc(&ep_buf_pool, block, cfg->ep_mps, K_MSEC(10))) {
 		LOG_ERR("Memory allocation time-out");
 		return -ENOMEM;
 	}
 
-	s_Device.eps[ep_abs_idx].ep_mps = cfg->ep_mps;
-	if (s_Device.eps[ep_abs_idx].ep_enabled) {
-		s_Device.interface->deviceControl(s_Device.controllerHandle, kUSB_DeviceControlEndpointDeinit, (void *)(&cfg->ep_addr));
```
[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply drivers/usb/device/usb_dc_mcux_ehci.c]: sorry, what do you mean?

**Comment by jfischer-phytec-iot** [comment]
Rebased. 

@MaureenHelm The header file modules/hal/nxp/mcux/middleware/usb/device/port/zephyr/usb_dc_mcux.h is a part of the driver. It hides a lot of things, has unnecessary complexity, and does not help to maintain the driver. This file should be moved to drivers/usb/device/usb_dc_mcux_hal.h and reworked/simplified. I would like to do that, this would also help to fix high speed support what I really would like to improve and test with imxrt1050.

File: drivers/usb/device/usb_dc_mcux_ehci.c [hunk 4]
```diff
@@ -45,65 +51,93 @@ K_MEM_POOL_DEFINE(ep_buf_pool, 16, 1024, EP_BUF_NUMOF_BLOCKS, 4);
...
 {
-	LOG_DBG("detached.");
-	if (s_Device.controllerHandle != NULL) {
-		s_Device.interface->deviceControl(s_Device.controllerHandle, kUSB_DeviceControlStop, NULL);
-		s_Device.interface->deviceDeinit(s_Device.controllerHandle);
-		s_Device.controllerHandle = NULL;
+	usb_status_t status;
+
+	if (dev_data.controllerHandle == NULL) {
+		LOG_WRN("Device not attached");
```
[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply drivers/usb/device/usb_dc_mcux_ehci.c]: That does not really matter, it is the current state of the device

File: drivers/usb/device/usb_dc_mcux_ehci.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (emob-nordic) [reply drivers/usb/device/usb_dc_mcux_ehci.c]: At first sight I was confused why You have removed
```
s_Device.interface->deviceControl(s_Device.controllerHandle, kUSB_DeviceControlEndpointDeinit ... );
```
But now I assume it has no sense to do that as later there is 
```
status = dev_data.interface->deviceControl(dev_data.controllerHandle,
			kUSB_DeviceControlEndpointInit, &ep_init);
```

I understand that You made some cleanup here.

**Review by eobalski — APPROVED** [review]

**Comment by jhedberg** [comment]
@MaureenHelm did you still want someone else to review this? I see you mentioned @MarkWangChinese back in December.

**Comment by jhedberg** [comment]
@jfischer-phytec-iot there seems to be a conflict with `drivers/usb/device/usb_dc_mcux_ehci.c` so you'll need to rebase. @MaureenHelm is someone else still expected to review this?

**Comment by hakehuang** [comment]
test pass in this PR @MaureenHelm , @dlech 
```
_descriptor_start 0x20002b5c
D: __usb_descriptor_end 0x20002bcd
D: Configuration descriptor 0x20002b6e
D: Interface descriptor 0x20002b77
D: Endpoint descriptor 0x20002b80
D: Fixing EP address 81 -> 81
D: endpoint 0x81
D: Now the wTotalLength is 25
D: idx_max 11, ascii_idx_max 5, buf 0x20002b8d
D: char R : 52, idx 5 -> 11
D: char Y : 59, idx 4 -> 9
D: char H : 48, idx 3 -> 7
D: char P : 50, idx 2 -> 5
D: char E : 45, idx 1 -> 3
D: char Z : 5a, idx 0 -> 1
D: idx_max 13, ascii_idx_max 6, buf 0x20002b9b
D: char V : 56, idx 6 -> 13
D: char E : 45, idx 5 -> 11
D: char D : 44, idx 4 -> 9
D: char - : 2d, idx 3 -> 7
D: char B : 42, idx 2 -> 5
D: char S : 53, idx 1 -> 3
D: char U : 55, idx 0 -> 1
D: Serial Number
D: d7 49 4c 5a 46 00 00 50 |.ILZF..P
D: idx_max 31, ascii_idx_max 15, buf 0x20002bab
D: char 7 : 37, idx 15 -> 31
D: char D : 44, idx 14 -> 29
D: char 9 : 39, idx 13 -> 27
D: char 4 : 34, idx 12 -> 25
D: char C : 43, idx 11 -> 23
D: char 4 : 34, idx 10 -> 21
D: char A : 41, idx 9 -> 19
D: char 5 : 35, idx 8 -> 17
D: char 6 : 36, idx 7 -> 15
D: char 4 : 34, idx 6 -> 13
D: char 0 : 30, idx 5 -> 11
D: char 0 : 30, idx 4 -> 9
D: char 0 : 30, idx 3 -> 7
D: char 0 : 30, idx 2 -> 5
D: char 0 : 30, idx 1 -> 3
D: char 5 : 35, idx 0 -> 1
*** Booting Zephyr OS version 2.2.0-rc1  ***
D: lock usb_enable_lock mutex
D: Attached
D: set cb, ep: 0x81
D: unlock usb_enable_lock mutex
Running test suite test_device
===================================================================
starting test - test_usb_dc_api_invalid
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address/direction
E: Wrong endpoint index/address/direction
E: Wrong endpoint index/address
E: Wrong endpoint index/address
PASS - test_usb_dc_api_invalid
===================================================================
starting test - test_usb_dc_api
PASS - test_usb_dc_api
===================================================================
starting test - test_usb_dc_api_read_write
E: Wrong endpoint index/address/direction
E: Wrong endpoint index/address
PASS - test_usb_dc_api_read_write
===================================================================
starting test - test_usb_dc_api_invalid
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address
E: Wrong endpoint index/address/direction
E: Wrong endpoint index/address/direction
E: Wrong endpoint index/address
E: Wrong endpoint index/address
PASS - test_usb_dc_api_invalid
===================================================================
starting test - test_usb_deconfig
PASS - test_usb_deconfig
===================================================================
starting test - test_usb_disable
W: Device not attached
PASS - test_usb_disable
===================================================================
Test suite test_device succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Comment by carlescufi** [comment]
@MaureenHelm this fixes a bug, has been opened for a while and is scheduled for this release. Could you review it so we can include it in the release?
cc @jhedberg

**Review by jhedberg — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**